---
title: "BAZ2B Hap cell line - VSN normalisation"
author: "Uwe Schwartz"
date: "10/07/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_knit$set(root.dir = "~/Analysis/14_Baz2b_Microarray/")
out.path<-"~/Analysis/14_Baz2b_Microarray/analysis/01_conventional/"
date.x<-Sys.Date()
```

## Dependencies

Microarray analysis was performed as recommended in the Bioconductor tutorial [An end to end workflow for differential gene expression using Affymetrix microarrays](https://bioconductor.org/packages/devel/workflows/vignettes/maEndToEnd/inst/doc/MA-Workflow.html) (`r date.x`). 

#### Install workflow
```{r install, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')
BiocManager::install("maEndToEnd")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("hugene20sttranscriptcluster.db")
```


#### Load packages
```{r libraries, message=FALSE}
suppressPackageStartupMessages({library("maEndToEnd")})
library(ggrepel)
library("hugene20sttranscriptcluster.db")
library(xtable)
library(ggplot2)
library(vsn)
library(Cairo)
library(Glimma)
```

##Import data

First an annotation file was created describing the source of RNA used for analysis

```{r annotation file}
# path to files
raw_data_dir<-"data/Microarray data 2018 baz2B Hap1/Raw Data/CEL files/"

anno.data<-data.frame(condition=c(rep(c("wt","d_Baz2B"),c(2,2))),
           replica=rep(c("R1","R2"),2),
           row.names = list.files(raw_data_dir))

SDRF <- AnnotatedDataFrame(anno.data)

anno.data
```

load microarray data and create an expression set object
```{r, message=FALSE, results='hide', cache=TRUE}
raw_data <- oligo::read.celfiles(filenames = file.path(raw_data_dir, 
                                                       row.names(anno.data)),
                                 verbose = FALSE, phenoData = SDRF)
```

```{r checkpoint1, eval=FALSE, echo=FALSE }
dir.create(paste(out.path,"obj",sep="/"))
save(raw_data, file=paste(out.path,"obj/raw.data.rda",sep="/"))
#load( file=paste(out.path,"obj/raw.data.rda",sep="/"))
```



## Exploratory data analysis

The PCA of the raw data was ploted

```{r PCA}
# get log2 transformed intensities
exp_raw <- log2(Biobase::exprs(raw_data))
#calculate PCA
PCA_raw <- prcomp(t(exp_raw), scale. = FALSE)

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

#assign features
dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                     condition = pData(raw_data)$condition,
                     replica = pData(raw_data)$replica,
                     name = rownames(pData(raw_data)))

#plot data
g<-ggplot(dataGG, aes(PC1, PC2, label=name))+
  geom_point(aes(shape = condition, colour = replica, size=3)) +
  ggtitle("log-transformed raw expression data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5))+
  coord_fixed(ratio = sd_ratio) +
  scale_shape_manual(values = c(4,15)) + 
  scale_color_manual(values = c("darkorange2", "dodgerblue4"))+
  theme_bw()+theme(aspect.ratio=4/3)

```


```{r plot PCA, echo=FALSE}
print(g+geom_label_repel(hjust = 0.5, size=2.2,label.size = NA, fill=NA))

```


```{r PCA raw data, eval=FALSE, echo=FALSE }
dir.create(paste(out.path,"plots_exploratory",sep="/"), showWarnings = F)
pdf(paste(out.path,"plots_exploratory/raw_data_PCA.pdf",sep="/"), width = 4, 
    height = 4)
    print(g+geom_label_repel(hjust = 0.5, size=2.2,label.size = NA, fill=NA))
dev.off()
```



Boxplots of log2-intensities of un-normalized samples 
```{r boxplot of raw data intensities, fig.height=10}
oligo::boxplot(raw_data, target = "core", las=2,pars=list(par(mar=c(11,4,4,2))),
               main = "Boxplot of log2-intensitites for the raw data")
```


```{r boxplot raw data, eval=FALSE, echo=FALSE }
pdf(paste(out.path,"plots_exploratory/raw_data_boxplot.pdf",sep="/"), width = 4.5,
    height = 8)
    oligo::boxplot(raw_data, target = "core", las=2,
                   pars=list(par(mar=c(11,4,4,2))),
               main = "Boxplot of log2-intensitites for the raw data")
dev.off()
```




Deviation of expression intensity from the median expression of each array.

```{r, fig.height=8}
palmieri_eset <- oligo::rma(raw_data, normalize = FALSE, target='core')
row_medians_assayData <- 
  Biobase::rowMedians(as.matrix(Biobase::exprs(palmieri_eset)))

RLE_data <- sweep(Biobase::exprs(palmieri_eset), 1, row_medians_assayData)

RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- 
  tidyr::gather(RLE_data, array, log2_expression_deviation)

g<-ggplot2::ggplot(RLE_data_gathered, aes(array,
                                          log2_expression_deviation)) + 
  geom_boxplot(outlier.shape = NA) +
  ylim(c(-1, 1))+theme_bw() + 
  theme(axis.text.x = element_text(colour = "aquamarine4", 
                                   angle = 60, size = 6.5, hjust = 1 ,
                                   face = "bold"))
g
```

```{r deviation from median raw data, eval=FALSE, echo=FALSE }
pdf(paste(out.path,"plots_exploratory/raw_data_deviation_from_median.pdf",sep="/"), 
    width = 4, height = 7)
    print(g)
dev.off()
```

##Normalization
The data was normalized using the rma function of the oligo package. RMA shares information across arrays and uses the versatile quantile normalization method that will make the array intensity distributions match.The quantile normalization algorithm used by RMA works by replacing values by the average of identically ranked (within a single chip) values across arrays. 


```{r normalization}
baz2b_del_norm <- oligo::rma(raw_data, target="core")
```


```{r checkpoint2, eval=FALSE, echo=FALSE }
save(baz2b_del_norm, file=paste(out.path,"obj/baz2b_del_norm.rda",sep="/"))

load(file=paste(out.path,"obj/baz2b_del_norm.rda",sep="/"))
```


## exploratory of normalized data

Deviation of expression intensity from the median expression of each array.

```{r after norm deviation, fig.height=8}
row_medians_assayData <- 
  Biobase::rowMedians(as.matrix(Biobase::exprs(baz2b_del_norm)))

RLE_data <- sweep(Biobase::exprs(baz2b_del_norm), 1, row_medians_assayData)

RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- 
  tidyr::gather(RLE_data, array, log2_expression_deviation)

g<-ggplot2::ggplot(RLE_data_gathered, aes(array,
                                          log2_expression_deviation)) + 
  geom_boxplot(outlier.shape = NA) +
  ylim(c(-1, 1))+theme_bw() + 
  theme(axis.text.x = element_text(colour = "aquamarine4", 
                                   angle = 60, size = 6.5, hjust = 1 ,
                                   face = "bold"))
g
```

```{r deviation from median norm data, eval=FALSE, echo=FALSE }
pdf(paste(out.path,"plots_exploratory/norm_data_deviation_from_median.pdf",sep="/"), 
    width = 4, height = 7)
    print(g)
dev.off()
```




PCA of normalized samples

```{r PCA norm data, fig.height=8}
baz2b_del_norm_exp <- Biobase::exprs(baz2b_del_norm)
PCA <- prcomp(t(baz2b_del_norm_exp), scale = FALSE)

percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                     condition = pData(baz2b_del_norm)$condition,
                     replica = pData(baz2b_del_norm)$replica,
                     name = rownames(pData(baz2b_del_norm)))

g<-ggplot(dataGG, aes(PC1, PC2, label=name))+
  geom_point(aes(shape = condition, colour = replica), size=3) +
  ggtitle("PCA plot of RPM normalized data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5))+
  coord_fixed(ratio = sd_ratio) +
  scale_shape_manual(values = c(4,15)) + 
  scale_color_manual(values = c("darkorange2", "dodgerblue4"))+
  theme_bw()+theme(aspect.ratio=4/3)

  print(g+geom_label_repel(hjust = 0.5, size=2,label.size = NA, fill=NA))
```




```{r PCA norm data pdf, eval=FALSE, echo=FALSE }
pdf(paste(out.path,"plots_exploratory/norm_data_PCA.pdf",sep="/"), width = 4, 
    height = 4)
    print(g+geom_label_repel(hjust = 0.5, size=2.2,label.size = NA, fill=NA))
dev.off()
```


Hierarchical clustering by distance


```{r dist cluster}
condition_names <- ifelse(str_detect(pData(baz2b_del_norm)$condition,
                                     "wt"), "wt", "d_Baz2B")

replica_names <- ifelse(str_detect(pData(baz2b_del_norm)$replica,
                                     "R1"), "R1", "R2")

annotation_for_heatmap <-data.frame(condition = condition_names,  replica = replica_names)
row.names(annotation_for_heatmap) <- row.names(pData(baz2b_del_norm))

## calcualte distances
dists <- as.matrix(dist(t(baz2b_del_norm_exp), method = "manhattan"))

rownames(dists) <- row.names(pData(baz2b_del_norm))
hmcol <- rev(colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255))
colnames(dists) <- NULL
diag(dists) <- NA

ann_colors <- list( condition= c(wt = "chartreuse4",d_Baz2B = "burlywood3"),
                    replica= c(R1 = "blue4", R2 = "cadetblue2"))

pheatmap(dists, col = (hmcol), 
         annotation_row = annotation_for_heatmap,
         annotation_colors = ann_colors,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                           max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering heatmap for the calibrated samples")
    
```

## intensity based filtering

probes with intensities below the cutoff of 4 in at least 2 samples were filtered

```{r filter intensities}
#get medians of IP
rma_medians <- rowMedians(Biobase::exprs(baz2b_del_norm))


  hist_res <- hist(rma_medians, 100, col = "cornsilk1", freq = FALSE, 
                 main = "Histogram of the median intensities", 
                 border = "antiquewhite4",
                 xlab = "Median intensities", xlim=c(0,14))
  abline(v = 3.5, col = "coral4", lwd = 2)
  
  
########### filtering intensities higher in at least 2 samples 
samples_cutoff <-2


idx_man_threshold <- apply(Biobase::exprs(baz2b_del_norm), 1,
                           function(x){
                             sum(x > 3.5) >= samples_cutoff})
# filterd probes
table(idx_man_threshold)

baz2b_del_manfiltered <- subset(baz2b_del_norm, idx_man_threshold)

```


```{r checkpoint3, eval=FALSE, echo=FALSE }
#save(baz2b_del_manfiltered, file=paste(out.path,"obj/baz2b_del_manfiltered.rda",sep="/"))


load(file=paste(out.path,"obj/baz2b_del_manfiltered.rda",sep="/"))

```


## get gene annotation

convert the probe IDs to human readable genes

```{r annotation conversion}
anno_baz2b <- AnnotationDbi::select(hugene20sttranscriptcluster.db,
                                       keys =(featureNames(baz2b_del_manfiltered)),
                                       columns = c("ENSEMBL"),
                                       keytype = "PROBEID")
# which Probe-IDs do not have an ENSEMBL accession
table(is.na(anno_baz2b$ENSEMBL))
no_ENSEMBL_ID<-anno_baz2b$PROBEID[(is.na(anno_baz2b$ENSEMBL))]

anno_baz2b <- subset(anno_baz2b, !is.na(ENSEMBL)) 


# get multimapping ids
anno_grouped <- dplyr::group_by(anno_baz2b, PROBEID)
anno_summarized <- 
  dplyr::summarize(anno_grouped, no_of_matches = n_distinct(ENSEMBL))


## merge all gene names of those with multimapping ids
anno_filtered <- filter(anno_summarized, no_of_matches > 1)
probe_stats <- anno_filtered 

#number of mult-ID probes and no ENSEMBL
nrow(probe_stats)

ids_to_exlude <- (featureNames(baz2b_del_manfiltered) %in% c(probe_stats$PROBEID,
                                                         no_ENSEMBL_ID))
table(ids_to_exlude)

## remove multimapper
baz2b_del_final <- subset(baz2b_del_manfiltered, !ids_to_exlude)
validObject(baz2b_del_final)

fData(baz2b_del_final)$PROBEID <- rownames(fData(baz2b_del_final))


fData(baz2b_del_final) <- left_join(fData(baz2b_del_final), anno_baz2b)
nrow(fData(baz2b_del_final))

rownames(fData(baz2b_del_final)) <- fData(baz2b_del_final)$PROBEID 
validObject(baz2b_del_final)



###### add gene names

anno_baz2b_genes <- AnnotationDbi::select(hugene20sttranscriptcluster.db,
                                       keys =(featureNames(baz2b_del_final)),
                                       columns = c("SYMBOL","GENENAME"),
                                       keytype = "PROBEID")

mx.genes<-match(fData(baz2b_del_final)$PROBEID , anno_baz2b_genes$PROBEID)

fData(baz2b_del_final)<-(data.frame(fData(baz2b_del_final),
                        anno_baz2b_genes[mx.genes,c("SYMBOL","GENENAME")]))
validObject(baz2b_del_final)
```


```{r checkpoint4, eval=FALSE, echo=FALSE }
#save(baz2b_del_final, file=paste(out.path,"obj/baz2b_del_final.rda",sep="/"))

load(file=paste(out.path,"obj/baz2b_del_final.rda",sep="/"))
```


## Differential analysis

To identify significant RNA expression changes in RNA Baz2b deleted cells a linear model was fit.
The main idea is to fit a linear model to the expression data for each gene. Empirical Bayes and other methods are used to borrow information across genes for the residual variance estimation leading to moderated t-statistics, and stabilizing the analysis for experiments with just a small number of arrays. Conceptually, the final per gene variance is a mix of a prior variance and the per gene variance.


```{r diff analysis, fig.height=8}
##desig of linear model
pData(baz2b_del_final)
design<-model.matrix(~0+factor(c(1,1,2,2)))
colnames(design)<-c( "wt", "d_Baz2b")
fit<-lmFit(baz2b_del_final, design)

contrast.matrix<-makeContrasts(d_Baz2b-wt  ,levels = design)
fit2<-contrasts.fit(fit, contrast.matrix)
fit2<-eBayes(fit2)

expFinal<-exprs(baz2b_del_final)

## d_Baz2b - wt
res.baz2b_del<-topTable(fit2, coef=1, number = Inf , adjust="BH")

# number of significant changes after correction for multiple testing
dt<-(decideTests(fit2,p.value = 0.05))
summary(dt)


#show the top hits
head(res.baz2b_del)

#add probe intensities
res.baz2b_del.ext<-cbind(res.baz2b_del,
                        expFinal[match(res.baz2b_del$PROBEID, rownames(expFinal)),])

## pvalue distribution
 hist(res.baz2b_del$P.Value, col = brewer.pal(3, name = "Set2")[2],
     main = "Baz2b_del", xlab = "p-values")



# Volcano plot highlighting the top 20 genes

volcanoplot(fit2, coef = 1, style = "p-value", highlight = 20,
            names = fit2$genes$SYMBOL,
            xlab = "Log2 Fold Change", ylab = NULL, pch=16, cex=0.35, main=
               "Baz2b_del")
### MA-plot

ggplot(res.baz2b_del.ext, aes(AveExpr,logFC))+geom_point()+
    geom_point(data = subset(res.baz2b_del.ext,  adj.P.Val<0.05), col="red")+
    theme_bw()

```



```{r Volcano plot and expression table, eval=FALSE, echo=FALSE }
pdf(paste(out.path,"results/volcano.pdf",sep="/"), width = 6, 
    height = 6)
    volcanoplot(fit2, coef = 1, style = "p-value", highlight = 20, 
                names = fit2$genes$SYMBOL,
            xlab = "Log2 Fold Change", ylab = NULL, pch=16, cex=0.35, main=
              "Baz2b_del")
dev.off()

write.table(res.baz2b_del.ext, 
    file=paste(out.path,"results/res.baz2b_del.txt",sep="/"), sep="\t",
            quote=F, row.names = F)


pdf(paste(out.path,"results/MA_labels.pdf",sep="/"), width = 8, 
    height = 8)
    g<-ggplot(res.baz2b_del.ext, aes(AveExpr,logFC))+geom_point()+
        geom_point(data = subset(res.baz2b_del.ext,  adj.P.Val<0.05), col="red")+
        geom_text_repel(data = subset(res.baz2b_del.ext,  adj.P.Val<0.05),
         aes(label = SYMBOL),size = 3,color="red",
         box.padding = unit(0.35, "lines"),
         point.padding = unit(0.3, "lines"))+
        theme_bw()
    print(g)
dev.off()


pdf(paste(out.path,"results/MA.pdf",sep="/"), width = 6, 
    height = 6)
    g<-ggplot(res.baz2b_del.ext, aes(AveExpr,logFC))+geom_point()+
        geom_point(data = subset(res.baz2b_del.ext,  adj.P.Val<0.05), col="red")+
        theme_bw()
    print(g)
dev.off()


#### Glimma

data(arraydata)
arrays <- arraydata$arrays
targets <- arraydata$targets


glMDPlot(fit2, status=dt, coef=1, main=colnames(fit2)[1],
    counts=expFinal, groups=baz2b_del_final$condition,
    side.ylab="Log-expression", side.main="ProbeID",
    samples=colnames(expFinal),
    display.columns = c("SYMBOL", "ENSEMBL", "logFC","Adj.PValue"),
    path = paste(out.path,"results/",sep="/"))

```

## Gene ontology (GO) enrichment analysis

We can now try to characterize the identified differentially expressed genes more in detail by performing a GO enrichment analysis.

For every differentially expressed gene, we try to find genes with similar expression with  genefinder. The genefinder function returns a list with two elements for each gene: one with the indices of the background genes found and one with the distances to the DE-genes:

```{r GO analysis, fig.height=8}
DE_genes<-subset(res.baz2b_del.ext, adj.P.Val < 0.05)$PROBEID

back_genes_idx <- genefilter::genefinder(baz2b_del_final, 
                                        as.character(DE_genes), 
                                        method = "manhattan", scale = "none")
back_genes_idx <- sapply(back_genes_idx, function(x)x$indices)
back_genes <- featureNames(baz2b_del_final)[back_genes_idx]
back_genes <- setdiff(back_genes, DE_genes)

multidensity(list(
        all = res.baz2b_del.ext[,"AveExpr"] ,
        fore =res.baz2b_del.ext[DE_genes , "AveExpr"],
        back = res.baz2b_del.ext[rownames(res.baz2b_del.ext) %in% back_genes, "AveExpr"]),
        col = c("#e46981", "#ae7ee2", "#a7ad4a"),
     xlab = "mean expression",
   main = "DE genes for background-matching")

## running topGO
gene_IDs <- rownames(res.baz2b_del.ext)
in_universe <- gene_IDs %in% c(DE_genes, back_genes)
in_selection <- gene_IDs %in% DE_genes 

all_genes <- in_selection[in_universe]
all_genes <- factor(as.integer(in_selection[in_universe]))
names(all_genes) <- gene_IDs[in_universe] 

top_GO_data <- new("topGOdata", ontology = "BP", allGenes = all_genes,
 nodeSize = 10, annot = annFUN.db, affyLib = "hugene20sttranscriptcluster.db")

result_top_GO_elim <- 
  runTest(top_GO_data, algorithm = "elim", statistic = "Fisher")
result_top_GO_classic <- 
  runTest(top_GO_data, algorithm = "classic", statistic = "Fisher")


res_top_GO <- GenTable(top_GO_data, Fisher.elim = result_top_GO_elim,
        Fisher.classic = result_top_GO_classic,
        orderBy = "Fisher.elim" , topNodes = 100)

genes_top_GO <- printGenes(top_GO_data, whichTerms = res_top_GO$GO.ID,
    chip = "hugene20sttranscriptcluster.db", geneCutOff = 1000)

res_top_GO$sig_genes <- sapply(genes_top_GO, function(x){
                str_c(paste0(x[x$'raw p-value' == 2, "Symbol.id"],";"), 
                      collapse = "")
    })




head(res_top_GO[,1:8], 20)


numb.sig<-sum(res_top_GO$Fisher.elim<0.01)
showSigOfNodes(top_GO_data, score(result_top_GO_elim), firstSigNodes = numb.sig,
               useInfo = 'def')

```

```{r GO save results, eval=FALSE, echo=FALSE }
dir.create(paste(out.path,"results/enrichment",sep="/"))

pdf(paste(out.path,"results/enrichment/GO_tree.pdf",sep="/"), width = 8, 
    height = 5)
    showSigOfNodes(top_GO_data, score(result_top_GO_elim), firstSigNodes = numb.sig,
               useInfo = 'def')
dev.off()

write.table(res_top_GO, 
    file=paste(out.path,"results/enrichment/GO_table.txt",sep="/"), sep="\t",
            quote=F, row.names = F)


```

## Pathway analysis

```{r reactome}
entrez_ids <- mapIds(hugene20sttranscriptcluster.db, 
      keys = rownames(res.baz2b_del), 
      keytype = "PROBEID",
      column = "ENTREZID")

reactome_enrich <- enrichPathway(gene = entrez_ids[DE_genes], 
                                universe = entrez_ids[c(DE_genes, 
                                                        back_genes)],
                                organism = "human",
                                pvalueCutoff = 0.05,
                                qvalueCutoff = 0.9,
                                readable = TRUE)


barplot(reactome_enrich)

```

```{r reactome save results, eval=FALSE, echo=FALSE }


pdf(paste(out.path,"results/enrichment/reactome_bar.pdf",sep="/"), width = 8, 
    height = 5)
    barplot(reactome_enrich)
dev.off()

write.table(as.data.frame(reactome_enrich), 
    file=paste(out.path,"results/enrichment/reactome_table.txt",sep="/"), sep="\t",
            quote=F, row.names = F)


```


##Session info
```{r info}
sessionInfo()


```

